### PR TITLE
chore(lint): remove jshint

### LINKED ISSUE #2582: Remove JSHint
Once #2578 lands and we're all happy w/ ESLint, we should remove the old JSHint subsystem (since we don't need redundant linting).

https://github.com/mozilla/fxa-content-server/search?utf8=%E2%9C%93&q=jshint

### PR DESCRIPTION
**Pull request by TDA**
Fixes #2582
Remove the evil JSHint from our code
All hail ESLint
Fixed README.md and CONTRIBUTING.md to remove JSHint and add ESLint
Fixed camelcase errors due to removing JSHint inline
@zaach 
@vladikoff 
@pdehaan 

ref #2578

### CODE REVIEW

File: app/scripts/lib/channels/mixins/postmessage_receiver.js [hunk 1]
```diff
@@ -1,12 +1,12 @@
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
-'use strict';
 
 /**
  * A channel mixin that expects responses to sent messages. Responses
  * are expected to be received via a postMessage to the window.
  */
+ 'use strict';
```
Comment (pdehaan) [inline app/scripts/lib/channels/mixins/postmessage_receiver.js]: Why this change?
And does moving the `'use strict'` beneath the comment break any JSDoc-type-functionality (if we're even using JSDoc for this repo)? /cc @vladikoff

File: app/scripts/lib/oauth-client.js [hunk 2]
```diff
@@ -39,8 +39,6 @@ function (xhr, OAuthErrors) {
 
     // params = { assertion, client_id, scope }
     getToken: function (params) {
-      /* jshint camelcase: false */
-
```
Comment (pdehaan) [inline app/scripts/lib/oauth-client.js]: This is interesting... Is ESLint not caring about camelcase things?
Should it? http://eslint.org/docs/rules/camelcase.html

File: app/tests/spec/lib/xss.js [hunk 3]
```diff
@@ -37,9 +37,7 @@ function (chai, _, XSS, Constants) {
       });
 
       it('disallows javascript: href', function () {
-        /*eslint-disable */
-        expectEmpty('javascript:alert(1)'); // jshint ignore:line 
-        /*eslint-enable */
+        expectEmpty('javascript:alert(1)'); //eslint-disable-line no-script-url
```
Comment (pdehaan) [inline app/tests/spec/lib/xss.js]: Nice!

**Comment by vladikoff** [comment]
forgot to remove jshint stuff from package.json :smile:

File: app/scripts/lib/channels/mixins/postmessage_receiver.js (hunk 1 above)
Comment (TDA) [inline app/scripts/lib/channels/mixins/postmessage_receiver.js]: Just for the sake of consistency, all the other files seem to have the use strict after the block comments, before the line comments.

**Comment by TDA** [comment]
@vladikoff aha! This is why I have you check everything :D you sir, are awesome! Am removing it right now ;)

File: app/scripts/lib/oauth-client.js (hunk 2 above)
Comment (TDA) [inline app/scripts/lib/oauth-client.js]: I guess it should, should I add it in?

**Comment by TDA** [comment]
Just realized I had to change the readme to include eslint, and remove jshint.

**Comment by pdehaan** [comment]
```sh
$ grep -irn "jshint" .
```
1. ./.gitignore:30:!.jshintrc
2. ./app/scripts/vendor/crosstab.js:389:        /*jshint bitwise: false*/
3. ./CHANGELOG.md:791:  * Update oauth client name, jshint ([566355fb](https://github.com/mozilla/fxa-content-server/commit/566355fbcb6514375e424bb790bf244f78bd04e1))
4. ./CHANGELOG.md:837:  * Fix the jshint errors and warnings in all of the tests. ([ad82aac6](https://github.com/mozilla/fxa-content-server/commit/ad82aac66cc2bb82c6bac3b52d7097b9133fdbf1))
5. ./CONTRIBUTING.md:21:  - Run `grunt jshint` to make sure your code passes linting.
6. ./CONTRIBUTING.md:29:- Run `grunt jshint` and `npm test` to make sure your code passes linting and tests.
7. ./CONTRIBUTING.md:78:| `grunt jshint` | run JSHint on client side and testing JavaScript. |
8. ./server/.jshintrc:2:  "extends": "../.jshintrc",

We can ignore the .gitignore weirdness, and the CHANGELOG.md files...
Probably ignore the _crosstab.js_ since it's in a /vendor/ dir (presumably 3rd party)
But also looks like we have a /server/.jshintrc file. We should remove that (and see if we need an .eslintrc file in that dir).

**Comment by TDA** [comment]
@pdehaan removed the server .jshintrc, i wonder how sublime text missed that :| grep it is from now :D

**Comment by TDA** [comment]
@vladikoff 
@pdehaan 
@zaach 
@shane-tomlinson 
ITS ONE WEEK! :D time to kill jshint! :"D rebase and kill? i mean merge :dancer:

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -18,7 +18,7 @@ Patches should be submitted as pull requests (PR).
 
 Before submitting a PR:
 - Your code must run and pass all the automated tests before you submit your PR for review. "Work in progress" pull requests are allowed to be submitted, but should be clearly labeled as such and should not be merged until all tests pass and the code has been reviewed. 
-  - Run `grunt jshint` to make sure your code passes linting.
+  - Run `grunt eslint` to make sure your code passes linting.
```
Comment (shane-tomlinson) [inline CONTRIBUTING.md]: Can you change `eslint` to `lint` here so all the linters are run?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -26,7 +26,7 @@ When submitting a PR:
 - You agree to license your code under the project's open source license ([MPL 2.0](/LICENSE)).
 - Base your branch off the current `master` (see below for an example workflow).
 - Add both your code and new tests if relevant.
-- Run `grunt jshint` and `npm test` to make sure your code passes linting and tests.
+- Run `grunt eslint` and `npm test` to make sure your code passes linting and tests.
```
Comment (shane-tomlinson) [inline CONTRIBUTING.md]: Same here about `eslint`=>`lint`

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -75,7 +75,7 @@ reference GitHub issues that this commit **Closes**.
 
 | Task | Description |
 |------|-------------|
-| `grunt jshint` | run JSHint on client side and testing JavaScript. |
+| `grunt eslint` | run ESLint on client side and testing JavaScript. |
```
Comment (shane-tomlinson) [inline CONTRIBUTING.md]: Maybe add a new section `grunt lint` that talks about runing all the linters.

We have a few missing lint related tasks from this section: `jsonlint`, `jscs` and `amdcheck`

File: app/scripts/lib/oauth-client.js (hunk 2 above)
Comment (shane-tomlinson) [inline app/scripts/lib/oauth-client.js]: yeah, please add camel case to eslint if it has the ability to handle that.

File: app/scripts/head/startup-styles.js [hunk 7]
```diff
@@ -8,8 +8,6 @@
 
 // NOTE: This file must run in IE8+, so no ES5/ES6 features!
 
-//jshint es3: true, es5: false, esnext: false
```
Comment (shane-tomlinson) [inline app/scripts/head/startup-styles.js]: Is there any way to tell eslint that the files in `head` must all use es3 only syntax?

File: app/scripts/lib/assertion.js [hunk 8]
```diff
@@ -24,14 +24,13 @@ function (P, jwcrypto) {
...
       if (window.crypto && window.crypto.getRandomValues) {
-        return P(true); /*jshint ignore:line*/
+        return P(true);
       }
     } catch(e) {
       // some browsers blow up when trying to query window.crypto.
     }
 
     // wah wah, we need to get entropy from the server.
-    //jshint validthis: true
```
Comment (shane-tomlinson) [inline app/scripts/lib/assertion.js]: Can eslint check for potentially invalid `this` usage?

File: app/scripts/lib/channels/duplex.js [hunk 9]
```diff
@@ -155,7 +155,7 @@ define([
      * @param {Object} [data]
      * @return {String}
      */
-    createMessageId: function (command, data) { //jshint ignore: line
+    createMessageId: function (command, data) {
```
Comment (shane-tomlinson) [inline app/scripts/lib/channels/duplex.js]: Does eslint check for unused variables? These are both unused in the traditional sense, but are used by the docs.

**Comment by shane-tomlinson** [comment]
On the whole, I like this. There are a few things @pdehaan and I have asked to see if eslint support:

* check for unused variables.
* check for potentially invalid `this` usage.
* specify some files must use es3 only syntax.

File: app/scripts/lib/oauth-client.js (hunk 2 above)
Comment (pdehaan) [inline app/scripts/lib/oauth-client.js]: @TDA, interesting, it looks like I disabled the `camelcase` config in the root `eslint-config-fxa` module (see https://github.com/pdehaan/eslint-config-fxa/blob/master/index.js#L9). We can enable it for the "fxa/client" template and set it to `2` (error), and when I try that locally it looks like we get 19 errors which we may need to inline ignore for those specific lines/blocks:

```sh
$ grunt eslint
Running "eslint:grunt" (eslint) task

grunttasks/clean.js
  7:4  error  Identifier 'tos_pp' is not in camel case  camelcase

grunttasks/copy.js
  25:4  error  Identifier 'tos_pp' is not in camel case       camelcase
  49:4  error  Identifier 'error_pages' is not in camel case  camelcase

grunttasks/l10n-generate-pages.js
  113:10  error  Identifier 'lang_dir' is not in camel case  camelcase

grunttasks/marked.js
  23:4  error  Identifier 'tos_pp' is not in camel case  camelcase

grunttasks/po2json.js
  17:6  error  Identifier 'output_filename' is not in camel case   camelcase
  45:6  error  Identifier 'output_transform' is not in camel case  camelcase

grunttasks/replace.js
  7:4  error  Identifier 'tos_pp' is not in camel case  camelcase

grunttasks/usemin.js
   7:4  error  Identifier 'client_rendered' is not in camel case  camelcase
  15:4  error  Identifier 'server_rendered' is not in camel case  camelcase

grunttasks/yeoman.js
  14:4  error  Identifier 'page_template_src' is not in camel case   camelcase
  15:4  error  Identifier 'page_template_dist' is not in camel case  camelcase
  16:4  error  Identifier 'strings_src' is not in camel case         camelcase
  17:4  error  Identifier 'strings_dist' is not in camel case        camelcase
  21:4  error  Identifier 'tos_pp_repo_dest' is not in camel case    camelcase
  22:4  error  Identifier 'tos_md_src' is not in camel case          camelcase
  23:4  error  Identifier 'tos_html_dest' is not in camel case       camelcase
  24:4  error  Identifier 'pp_md_src' is not in camel case           camelcase
  25:4  error  Identifier 'pp_html_dest' is not in camel case        camelcase

✖ 19 problems (19 errors, 0 warnings)

Warning: Task "eslint:grunt" failed. Use --force to continue.

Aborted due to warnings.
```

File: app/scripts/lib/assertion.js (hunk 8 above)
Comment (pdehaan) [inline app/scripts/lib/assertion.js]: Closest answer I've found is https://github.com/eslint/eslint/issues/1636

File: CONTRIBUTING.md (hunk 4 above)
Comment (TDA) [inline CONTRIBUTING.md]: okie done!

File: CONTRIBUTING.md (hunk 5 above)
Comment (TDA) [inline CONTRIBUTING.md]: done!

File: app/scripts/head/startup-styles.js (hunk 7 above)
Comment (TDA) [inline app/scripts/head/startup-styles.js]: There doesnt seem to be a way for this. There is a switch for es5/6 but not for es3.

File: app/scripts/lib/assertion.js (hunk 8 above)
Comment (TDA) [inline app/scripts/lib/assertion.js]: @pdehaan strict-this hasnt been implemented in ESLint though, has it?

File: app/scripts/lib/channels/duplex.js (hunk 9 above)
Comment (TDA) [inline app/scripts/lib/channels/duplex.js]: I think we already are checking for unused-variables, the rule "no-unused-vars" is set to [2, {"vars": "all", "args": "none"}], which should theoretically throw an error unless the variable was 'read' somewhere. In this file, the variables are in fact read in the following lines, where they are assigned to the obj-literal, so even with turning "args" on, we wont be getting an error i guess. Do you want me to change the "args" parameter to "all"?
@shane-tomlinson

File: app/scripts/lib/assertion.js (hunk 8 above)
Comment (pdehaan) [inline app/scripts/lib/assertion.js]: @TDA I don't think so. I can't find that rule in the [CHANGELOG](https://github.com/eslint/eslint/blob/master/CHANGELOG.md) or on the [RULES](http://eslint.org/docs/rules/) page, only in that https://github.com/mozilla/fxa-content-server/pull/2597#discussion_r32895193 above.

Comment (TDA) [inline app/scripts/lib/assertion.js]: So, just let it be?

Comment (pdehaan) [inline app/scripts/lib/assertion.js]: I don't think there is anything we can do here until ESLint implements https://github.com/eslint/eslint/issues/1636

File: app/scripts/lib/oauth-client.js (hunk 2 above)
Comment (pdehaan) [inline app/scripts/lib/oauth-client.js]: @TDA: Let's add `camelcase: 2` to the base /.eslintrc file and tweak the 19 places where we have errors and disable the rule on those specific lines.

Comment (TDA) [inline app/scripts/lib/oauth-client.js]: Updated this to use the fxa/client

**Comment by TDA** [comment]
@pdehaan Its done. :)

File: app/tests/spec/lib/app-start.js [hunk 10]
```diff
@@ -37,6 +37,7 @@ function (chai, sinon, AppStart, Session, Constants, p, Url, OAuthErrors,
   'use strict';
 
   /*global describe, beforeEach, it*/
```
Comment (pdehaan) [inline app/tests/spec/lib/app-start.js]: @TDA: IIUC, since we're setting an ESLint `env` of "mocha", we probably don't need to declare these inline `global` vars anymore.

You may want to do a quick `grep` and see if we can remove the comments throughout the */tests/* directories.

File: app/tests/spec/lib/oauth-client.js [hunk 11]
```diff
@@ -17,7 +17,7 @@ function (chai, sinon, OAuthClient, OAuthErrors, Xhr, p) {
   'use strict';
 
   /*global beforeEach, afterEach, describe, it*/
```
Comment (pdehaan) [inline app/tests/spec/lib/oauth-client.js]: @TDA: See previous comment about possibly removing these `/*global ... */` blocks.

**Comment by pdehaan** [comment]
Re: @shane-tomlinson's https://github.com/mozilla/fxa-content-server/pull/2597#issuecomment-113884986,

> 1. check for unused variables.
>> @pdehaan: This should be done and working already. The current ESLint config should fail on unused variables, but ignore unused function arguments (similar to what JSHint was doing previously, IIRC).

> 2. check for potentially invalid this usage.
>> @pdehaan: Not currently supported in ESLint, sadly. See https://github.com/eslint/eslint/issues/1636

> 3. specify some files must use es3 only syntax.
>> @pdehaan: I don't think this is [currently] supported by ESLint. I'll file an ESLint bug shortly.

File: app/scripts/lib/channels/duplex.js (hunk 9 above)
Comment (hzoo) [inline app/scripts/lib/channels/duplex.js]: @shane-tomlinson Yeah it checks for unused variables but won't count jsdoc usage as valid https://github.com/eslint/eslint/issues/2813

Comment (TDA) [inline app/scripts/lib/channels/duplex.js]: @hzoo Thanks for the heads up! :)

**Comment by pdehaan** [comment]
Alright, Travis is green, so I'm merging this since I think it's a blocker for your other PR.
:+1:

File: app/scripts/lib/assertion.js (hunk 8 above)
Comment (IanVS) [inline app/scripts/lib/assertion.js]: As a note, this has been added in `v1.0.0` of ESLint as [no-invalid-this](http://eslint.org/docs/1.0.0/rules/no-invalid-this)

Comment (TDA) [inline app/scripts/lib/assertion.js]: @pdehaan Would you be adding that to the config-fxa? @IanVS Thank you for the heads up :)

Comment (pdehaan) [inline app/scripts/lib/assertion.js]: @IanVS @TDA Great tip, thanks for the heads up. I still haven't looked at the [breaking] changes in **eslint@1** (versus our current v0.24.1), but yeah, I'll try and normalize all these in the eslint-config-fxa module, plus make sure we still have our solid set of default rules since I think the new ESLint 1.x will reset rules by default. Then once ESLint v1 is final (and the related grunt-eslint is updated), we can update our ~10 repos.

Ah, looks like I need to read this today: http://eslint.org/docs/1.0.0/user-guide/migrating-to-1.0.0
